src/time_integrator.f90

```diff
@@ -5,73 +5,270 @@ module m_time_integrator
       end do
     end do
 
-  end function constructor
+  end function init
 
   subroutine step(self, u, v, w, du, dv, dw, dt)
     implicit none
 
-    class(time_intg_t), intent(in) :: self
-    class(field_t), intent(inout) :: u, v, w
-    class(field_t), intent(in) :: du, dv, dw
+    class(time_intg_t), intent(inout) :: self
+    class(field_t), target, intent(inout) :: u, v, w
+    class(field_t), target, intent(in) :: du, dv, dw
 
     real(dp), intent(in) :: dt
 
-    call self%backend%vecadd(dt, du, 1._dp, u)
-    call self%backend%vecadd(dt, dv, 1._dp, v)
-    call self%backend%vecadd(dt, dw, 1._dp, w)
+    ! assign pointer to variables
+    self%curr(1)%ptr => u
+    self%curr(2)%ptr => v
+    self%curr(3)%ptr => w
+
+    ! assign pointer to variables
+    self%deriv(1)%ptr => du
+    self%deriv(2)%ptr => dv
+    self%deriv(3)%ptr => dw
+
+    if (self%method < 5) then
+      call self%adams_bashforth(dt)
+    else
+      call self%runge_kutta(dt)
+    end if
 
   end subroutine step
 
-  subroutine adams_bashford_1st(vels, olds, coeffs)
-    type(flist_t) :: vels(:), olds(:)
-    real :: coeffs(:)
+  subroutine runge_kutta(self, dt)
+    class(time_intg_t), intent(inout) :: self
+    real(dp), intent(in) :: dt
+
+    integer :: i, j
+
+    ! update solution
+    if (self%istage == self%nstage) then
+      do i = 1, self%nvars
+        ! update step solution from stage derivative
+        if (self%nstage > 1) then
+          call self%backend%vecadd(1.0_dp, self%olds(i, 1)%ptr, 0._dp, &
+                                   self%curr(i)%ptr)
+        end if
+
+        do j = 1, self%nstage - 1
+          call self%backend%vecadd(self%rk_b(j, self%nstage)*dt, &
+                                   self%olds(i, j + 1)%ptr, &
+                                   1._dp, self%curr(i)%ptr)
+        end do
+        call self%backend%vecadd(self%rk_b(self%nstage, self%nstage)*dt, &
+                                 self%deriv(i)%ptr, &
+                                 1._dp, self%curr(i)%ptr)
+
+      end do
+
+      ! reset stage counter
+      self%istage = 1
+    else
+      do i = 1, self%nvars
+        ! save step initial condition
+        if (self%istage == 1) then
+          call self%backend%vecadd(1.0_dp, self%curr(i)%ptr, 0._dp, &
+                                   self%olds(i, 1)%ptr)
+        end if
+
+        ! save stage derivative
+        call self%backend%vecadd(1.0_dp, self%deriv(i)%ptr, 0._dp, &
+                                 self%olds(i, self%istage + 1)%ptr)
+
+        ! update stage solution
+        if (self%istage > 1) then
+          call self%backend%vecadd(1.0_dp, self%olds(i, 1)%ptr, 0._dp, &
+                                   self%curr(i)%ptr)
+        end if
+        do j = 1, self%istage
+          call self%backend%vecadd(self%rk_a(j, self%istage, self%nstage)*dt, &
+                                   self%olds(i, j + 1)%ptr, &
+                                   1._dp, self%curr(i)%ptr)
+        end do
```

Comment:
This loop can be eliminated if we decide to only consider "diagonal" explicit RK. For the current implementation, we are simply multiplying arrays by zero for `j < self%istage`.